nixos/upower: Allow customization of UPower.conf

mat8913 opened the pull request:
Documentation of options and default values taken from the upstream
UPower.conf. The documentation was modified slightly to make more sense
when displayed on https://nixos.org/nixos/options.html.

A copy of upstream UPower.conf can be found here:
https://cgit.freedesktop.org/upower/tree/etc/UPower.conf?id=28bd86c181e2510ef6a1dc7cfa26f97803698a79

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

On my laptop, the battery jumps from 20% straight down to 0%, not giving it enough time to hibernate under the default settings.

#25815

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I tested this on my laptop running NixOS 19.09 by cherry-picking the commit onto the release branch and adding this to my config:

```
  services.upower.percentageLow = 35;
  services.upower.percentageCritical = 30;
  services.upower.percentageAction = 25;
```

###### Notify maintainers

cc @mmahut @worldofpeace

mat8913 committed:
nixos/upower: Allow customization of UPower.conf

Documentation of options and default values taken from the upstream
UPower.conf. The documentation was modified slightly to make more sense
when displayed on https://nixos.org/nixos/options.html.

A copy of upstream UPower.conf can be found here:
https://cgit.freedesktop.org/upower/tree/etc/UPower.conf?id=28bd86c181e2510ef6a1dc7cfa26f97803698a79

filalex77 commented on nixos/modules/services/hardware/upower.nix:
```diff
@@ -54,6 +220,24 @@ in
+    systemd.services.upower.environment.UPOWER_CONF_FILE_NAME = pkgs.writeTextFile {
+    name = "UPower.conf";
+    text = ''
+      [UPower]
```
How about using `lib.generators.toINI`? You can keep the options separately for documentation:

```
text = generators.toINI {} {
  EnableWattsUpPro = cfg.enableWattsUpPro;
  # ...
};
```

`toINI` has sensible defaults, e.g. converting booleans to proper `"true"` and `"false"` strings.

mat8913 replied:
Thanks for the review. I didn't know about `toINI`, trying it now.

matthewbauer commented on nixos/modules/services/hardware/upower.nix:
```diff
@@ -54,6 +220,24 @@ in
 
     systemd.packages = [ cfg.package ];
 
+    systemd.services.upower.environment.UPOWER_CONF_FILE_NAME = pkgs.writeTextFile {
```
We have to be careful about this because /etc/UPower/UPower.conf could be used in some config settings (via ```environment.etc."UPower/UPower.conf"```.)

worldofpeace replied:
The alternative would be to create the file at this location, which would be more convenient if people wanted to inspect this result.

However either way this should have a release note to notify people of the new module and to use it to manager upower's configuration.

mat8913 replied:
Should I change it to this?

```
systemd.services.upower.environment.UPOWER_CONF_FILE_NAME = "/etc/UPower/UPower.conf";
environment.etc."UPower/UPower.conf".text = generators.toINI {} {
  UPower = {
    EnableWattsUpPro = cfg.enableWattsUpPro;
    NoPollBatteries = cfg.noPollBatteries;
    IgnoreLid = cfg.ignoreLid;
    UsePercentageForPolicy = cfg.usePercentageForPolicy;
    PercentageLow = cfg.percentageLow;
    PercentageCritical = cfg.percentageCritical;
    PercentageAction = cfg.percentageAction;
    TimeLow = cfg.timeLow;
    TimeCritical = cfg.timeCritical;
    TimeAction = cfg.timeAction;
    CriticalPowerAction = cfg.criticalPowerAction;
  };
};
```

worldofpeace replied:
I have a stronger preference for having the file only at `/etc`.
It seems `UPOWER_CONF_FILE_NAME` was implemented for their integration testing.

Also, did you need to set the file name to that path manually?
It should be compiled it to read the configuration from this directory.

mat8913 replied:
On NixOS, UPower loads its config from its store path by default. (`/nix/store/wr1699ff89pp60fcfbv42ixw8ci8h9m9-upower-0.99.11/etc/UPower/UPower.conf` on my computer). That's why I needed to set `UPOWER_CONF_FILE_NAME`.

worldofpeace reviewed (changes requested):
See @matthewbauer and my comments.

A release note can be added at
* https://github.com/NixOS/nixpkgs/blob/099e7759f100b366a389e6c92709d57ede81b86e/nixos/doc/manual/release-notes/rl-2003.xml#L69

mat8913 committed:
upower: load config from /etc

In the process of making UPower.conf customizable (#73968), it came up
that UPower doesn't load its config from /etc by default.

The UPower derivation is modified to make it load its config from /etc
at runtime, but still install the default config to its nix store path
as before.

The UPower module is modified to put the config in /etc.

worldofpeace commented:
@mat8913 Could you take the diff that touches the upower package at https://github.com/NixOS/nixpkgs/pull/74903/commits/56b4d68f92e9e3120b040855e255dbba57ebdbdb and apply it here?

That way you don't have to make it load the config from the package in that PR, then rebase this and undo this.

worldofpeace commented on nixos/doc/manual/release-notes/rl-2003.xml:
```diff
@@ -77,6 +77,12 @@
+   <listitem>
+    <para>
+    UPower's configuration is now managed by NixOS and can be customized
+    via <literal>services.upower.*</literal>.
```
```suggestion
    via <option>services.upower</option>.
```

mat8913 commented:
OK, I think I've got this right. I'll close the other PR.

worldofpeace commented:
Thanks @mat8913, upower is working fine on my machine with these patches.
